app/models/sources/linode.rb

```diff
@@ -0,0 +1,190 @@
+        # If the id is not found then adds to the source's persistent error log
+        # Parameters:
+        #   None
+        # Returns:
+        #   - a numeric id representing the image on Linode
+        #   - nil if not found
+        def retrieve_image_id
+          name = config['image_name']
+
+          img_data = connection.image_list.body["DATA"].find { |i| i["LABEL"] == name }
+          if img_data.nil?
+            add_error "There is no image named \#{name}."
+          else
+            id = img_data["IMAGEID"]
+          end
+          id
+        end
+
+        # retrieve_flavor_id()
+        # retrieve_kernel_id()
+        # retrieve_data_center_id()
+        # Retrieve flavor/kernel/data center id from Linode
+        # If the id is not found then adds to the source's persistent error log
+        # Parameters:
+        #   None
+        # Returns:
+        #   - a numeric id representing the flavor/kernel/data center on Linode
+        #   - nil if not found
+        def retrieve_#{id_type}_id
+          name = config['#{id_type}_name']
+
+          id = connection.#{id_type.pluralize}.select do |i|
+            if '#{id_type}' == 'data_center' then i.location == name
+            else i.name == name end
+          end.first.andand.id
+          if id.nil?
+            names = connection.#{id_type.pluralize}.map(&:name).join(", ")
+            add_error "There is no #{id_type} named \#{name}. " \
+              "Options are: \#{names}"
+          end
+          id
+        end
+      RUBY
+    end
+
+    # names_to_ids()
+    # Retrieve all ids from the server and save them in the database
+    # Does not save to the database if at least one id is nil.
+    # Parameters:
+    #   None
+    def names_to_ids
+      config['image_id'] = retrieve_image_id
+      config['flavor_id'] = retrieve_flavor_id
+      config['kernel_id'] = retrieve_kernel_id
+      config['data_center_id'] = retrieve_data_center_id
+      unless config['image_id'].nil? \
+        || config['flavor_id'].nil? \
+        || config['kernel_id'].nil? \
+        || config['data_center_id'].nil?
+
+        self.save
+      end
+    end
+
+    def parse_proxy_name(name)
+      name_pattern = /^(\d+)-(\d+)-(\d+)-(\d+)-/
+      parsed_data = name_pattern.match(name)
+      if (!parsed_data.nil?)
+        server_details = {
+          :data_center_id => parsed_data.captures[0].to_i,
+          :flavor_id => parsed_data.captures[1].to_i,
+          :kernel_id => parsed_data.captures[2].to_i,
+          :image_id => parsed_data.captures[3].to_i
+        }
+      end
+      server_details
+    end
+
+    def create_server
+      # Error message says they accept names up to 50 chars...but they don't
+      # Changed uniqueness to be PID/timestamp since Linode eventually
+      # truncates the name futher in their system and UUID kept getting truncated.
+      name = "#{data_center_id}-#{flavor_id}-#{kernel_id}-#{image_id}-#{Process.pid}-#{Time.new.to_i}"[0,45]
```

Comment:
This method is only called from the single-threaded worker that handles events for this Source class. `Process.pid` will just make the name longer for no reason.